### DISCUSSION TITLE
Make pyarmor work with zipapp

### DISCUSSION THREAD

**Discussion by EmberLightVFX** [dashingsoft/pyarmor#1510] [Ideas]
I'm working on a module that I would need to package down into one single file. Using zipapp would be the best solution but zipapp doesn't work with .pyd files. Shiv and equal doesn't work as those can't be loaded as modules. zipapps can.

My idea is to modify the `__init__.py` file in the `pyarmor_runtime_006099` folder so the file itself includes the binaries (converted to base64) of the original `__init__.py` and `pyarmor_runtime.pyd`. When __init__.py gets execited it will create the init and runtime files in a temp-folder, add the temp-folder to sys using `sys.path.append(temp_dir)` and the import it like this `from pyarmor_runtime_006099 import __pyarmor__`

The system works in my tests with the built pyarmor files (with modified __init__.py) but if I obfuscate the code with `--assert-call` or `--assert-import` I get an protection exception after packing in into a zipapp.

So my feature request would be to be able to pack the obfuscation into a zipapp with `--assert-call` and `--assert-import` applied but ignored for the __init__.py file so it can genenrate the needed .pyd file and import it outside the zipapp.

I'm attaching my test-project here so you can try it out yourself. These are the steps needed:

1. Run `build_pyarmor.py` to generate the needed pyarmor code
2. Run `generate_init.py` and select `./build/aesl/pyarmor_runtime_006099/__init__.py` or `pyarmor_runtime.pyd`
  The script will fetch them both, take the content of `__init__template.py`, replace the needed parts and overwrite the old __init__.py file.
3. Run `build_zipapp.py` to generate the zipapp.
4. In `./dist` you should now have `demo.py` and `aesl.pyz`. Run demo.py to test it out
5. To make sure the non-zipapp version works with the modified __init__.py, copy the demo.py file from root into `.build` and run it.

[test.zip](https://github.com/dashingsoft/pyarmor/files/12749841/test.zip)

**Comment by jondy** [comment]
Refer to
https://pyarmor.readthedocs.io/en/latest/tutorial/advanced.html#filter-assert-function-and-import

**Reply by EmberLightVFX** [reply]
I disabled `--assert-call` and ran only with `--assert-import` in my tests now.
I have tried assert.import:excludes with any assert with all module names I can think of
```
"zipimport",
"aesl",
"importlib",
"_bootstrap",
"importlib._bootstrap",
"frozen",
"my_module",
"pyarmor_runtime_006099",
"pyarmor_runtime",
"__init__",
"__pyarmor__",
```

but it doesn't work.
I also tried a bunch of different excludes calls with all of the names above just in case:

```
assert_import:excludes=
assert-import:excludes=
assert.import:excludes=
assertimport:excludes=
assert:excludes=
import:excludes=
```

but still nothing... Are there anything I'm missing? Are there some extra protection on the pyarmor_runtime itself with --assert_import?

I'm attaching a small updated version of my test here: [test2.zip](https://github.com/dashingsoft/pyarmor/files/12769347/test2.zip)
It includes a `ignore_assert` list and the adding of `assert_import:excludes=` to the conf

**Comment by jondy** [comment]
If auto filter doesn't work, try to search `Patching source by inline marker` in the documentation, and use inline marker to check imported module or method manually.

**Reply by EmberLightVFX** [reply]
The problem lies in the files PyArmor generates and not in my source files, so I have no control over those files. At least not from what I find in the documentation.
Maybe that's the reason why `assert.import:excludes` doesn't solve it for me as the functions to exclude isn't part of my source code?
